magicblock-chainlink/src/remote_account_provider/chain_laser_actor.rs

```diff
@@ -0,0 +1,631 @@
+use std::{
+    collections::{HashMap, HashSet},
+    pin::Pin,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+use futures_util::{Stream, StreamExt};
+use helius_laserstream::{
+    client,
+    grpc::{
+        subscribe_update::UpdateOneof, CommitmentLevel, SubscribeRequest,
+        SubscribeRequestFilterAccounts, SubscribeUpdate,
+    },
+    ChannelOptions, LaserstreamConfig, LaserstreamError,
+};
+use log::*;
+use solana_account::Account;
+use solana_commitment_config::CommitmentLevel as SolanaCommitmentLevel;
+use solana_pubkey::Pubkey;
+use solana_sdk_ids::sysvar::clock;
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::StreamMap;
+
+use crate::remote_account_provider::{
+    pubsub_common::{
+        ChainPubsubActorMessage, MESSAGE_CHANNEL_SIZE,
+        SUBSCRIPTION_UPDATE_CHANNEL_SIZE,
+    },
+    RemoteAccountProviderError, RemoteAccountProviderResult,
+    SubscriptionUpdate,
+};
+
+type LaserResult = Result<SubscribeUpdate, LaserstreamError>;
+type LaserStreamUpdate = (usize, LaserResult);
+type LaserStream = Pin<Box<dyn Stream<Item = LaserResult> + Send>>;
+
+const PER_STREAM_SUBSCRIPTION_LIMIT: usize = 1_000;
+const SUBSCRIPTION_ACTIVATION_INTERVAL_MILLIS: u64 = 10_000;
+const SLOTS_BETWEEN_ACTIVATIONS: u64 =
+    SUBSCRIPTION_ACTIVATION_INTERVAL_MILLIS / 400;
+
+// -----------------
+// ChainLaserActor
+// -----------------
+/// ChainLaserActor manages gRPC subscriptions to Helius Laser or Triton endpoints.
```

Comment:
General comment: while Helius and Triton are currently the providers we use, we should be more agnostic, where possible, in the codebase and support all RPCs that are compatible with the Helius and Triton SDKs or interfaces. In practice this does not change anything, but naming and comments could be made more general, for example since multiple providers support Yellowstone gRPC.